### ISSUE TITLE
Regression in Implicit Search

### ISSUE THREAD

**Issue by adamgfraser** [lampepfl/dotty#11294]
## Compiler version

`3.0.0-M3`

## Minimized code

```Scala
object Reproducer extends App {

  trait ZPure[+W, -S1, +S2, -R, +E, +A]

  object ZPure {

    implicit final class UnifiedSyntax[W, S, R, E, A](private val self: ZPure[W, S, S, R, E, A]) extends AnyVal {
      def zipPar[B](that: ZPure[W, S, S, R, E, B]): ZPure[W, S, S, R, E, (A, B)] = ???
    }
  }
  
  trait CommutativeBoth[F[+_]] {
    def both[A ,B](fa: F[A], fb: F[B]): F[(A, B)]
  }
  
  implicit class CommutativeBothSyntax[F[+_], A](fa: => F[A]) {
    def zipPar[B](that: => F[B])(implicit both: CommutativeBoth[F]): F[(A, B)] = ???
  }

  object Example {

    lazy val foo: ZPure[Nothing, Unit, Unit, Any, String, Int] = ???
    foo zipPar foo
  }
  
  println("It compiles!")
}
```

## Output

```scala
no implicit argument of type Reproducer.CommutativeBoth[
  ([A] =>> Reproducer.ZPure[Nothing, Unit, Unit, Any, String, A])
] was found for parameter both of method zipPar in class CommutativeBothSyntax
```

## Expectation

This code compiles on Scala 2. It looks like the Scala 3 compiler is looking for a way to resolve the `zipPar` argument and finding `CommutativeBothSyntax` as one potential solution. However, it then determines there is no implicit `CommutativeBoth` instance and fails the compilation. Instead it should backtrack to the other potential solution in scope, `UnifiedSyntax`, and find that is indeed a valid solution.

**Comment by prolativ** [comment]
I think this is related to https://github.com/lampepfl/dotty/issues/7069. It's caused by the fact that in scala3 more deeply nested implicits have higher priority. 
As a workaround you could put `UnifiedSyntax` directly in `Reproducer` instead of in `Pure` (or anywhere else so that `UnifiedSyntax` and `CommutativeBothSyntax` are seen by the compiler with the same priority) and then the conflict in implicit search will make the compiler check both conversions

**Comment by adamgfraser** [comment]
@prolativ Thanks for the response!

I'm not sure that solution works for me here. I specifically want `UnifiedSyntax` to be in the `ZPure` companion object so it is automatically in implicit for anyone working with the `ZPure` data type. If I moved it somewhere else my users would need at least some import for the implicit to be in scope. I tried adding "nesting" by moving `UnifiedSyntax` to a low priority trait that the `ZPure` companion object extended but I don't think that is the kind of nesting you are referring to and that did not resolve the issue.

**Comment by prolativ** [comment]
@adamgfraser I believe this should work if you move both `UnifiedSyntax` and `CommutativeBothSyntax` to a trait extended by `ZPure` object.

**Comment by adamgfraser** [comment]
@prolativ Hmmm, that does work but I'm not sure if it is a solution more generally.  `CommutativeBothSyntax`needs to be available for a lot of other data types so we can't' just put it in the companion object for this one data type.

**Comment by prolativ** [comment]
@adamgfraser it looks like there's a simpler solution. The problem goes away if you take `Example` out of `Reproducer`:
```scala
object Reproducer extends App {

  trait ZPure[+W, -S1, +S2, -R, +E, +A]

  object ZPure {

    implicit final class UnifiedSyntax[W, S, R, E, A](private val self: ZPure[W, S, S, R, E, A]) extends AnyVal {
      def zipPar[B](that: ZPure[W, S, S, R, E, B]): ZPure[W, S, S, R, E, (A, B)] = ???
    }
  }
  
  trait CommutativeBoth[F[+_]] {
    def both[A ,B](fa: F[A], fb: F[B]): F[(A, B)]
  }
  
  implicit class CommutativeBothSyntax[F[+_], A](fa: => F[A]) {
    def zipPar[B](that: => F[B])(implicit both: CommutativeBoth[F]): F[(A, B)] = ???
  }
  
  println("It compiles!")
}

object Example {
  lazy val foo: Reproducer.ZPure[Nothing, Unit, Unit, Any, String, Int] = ???
  foo zipPar foo
}
```

**Comment by adamgfraser** [comment]
But isn't that just because we're not bringing the other potential implicit into scope at all? To clarify the example slightly, this does not compile:

```scala
object MyLibrary {

  trait ZPure[+W, -S1, +S2, -R, +E, +A]

  object ZPure {

    implicit final class UnifiedSyntax[W, S, R, E, A](private val self: ZPure[W, S, S, R, E, A]) extends AnyVal {
      def zipPar[B](that: ZPure[W, S, S, R, E, B]): ZPure[W, S, S, R, E, (A, B)] = ???
    }
  }
  
  trait CommutativeBoth[F[+_]] {
    def both[A ,B](fa: F[A], fb: F[B]): F[(A, B)]
  }
  
  implicit class CommutativeBothSyntax[F[+_], A](fa: => F[A]) {
    def zipPar[B](that: => F[B])(implicit both: CommutativeBoth[F]): F[(A, B)] = ???
  }
}

object UserCode extends App {
  import MyLibrary._
  
  lazy val foo: ZPure[Nothing, Unit, Unit, Any, String, Int] = ???
  foo zipPar foo
  
  println("It compiles!")
}
```

I want the user to do `import MyLibrary._` to bring into scope all the syntax and other functionality I have for them. But by doing this they trigger the issue with this specific implicit for this data type being unavailable to them.

**Comment by prolativ** [comment]
Ok, you're right, that worked because `CommutativeBothSyntax` was not in scope inside of `Example`. On the other hand if you expect the users to use a wildcard import from the library then my initial suggestion of moving `UnifiedSyntax` one level up should work. Alternatively can't you define `zipPar` on `ZPure` as a regular method instead of an extension?
And as I said at the beginning: these are workarounds. The best solution would be to fix the bug but I tried that some time ago and had to give up. I hope that after https://github.com/lampepfl/dotty/pull/11274/files got merged the fix might be slightly easier but I can't promise I'll have time to have a look at that again in the nearest future.

**Comment by adamgfraser** [comment]
Yes. I think the ideal solution would be a version of the code as originally written because then:

1. If the user does a wildcard import they get all the implicit syntax through type classes like `CommutativeBoth`
2. Even if the user doesn't do any imports at least extension methods defined on concrete data types work for them

Your proposed solution of moving everything to the same level is basically what we did (in our case relying solely on the `CommutativeBothSyntax`) but that means we get (1) not (2) so it is a second best solution.

Unfortunately we can't define `zipPar` directly as an operator on `ZPure` because it is only well defined when the `S1` and `S2` types are unified and at least in our experience we haven't found a good way to express that constraint without implicit syntax and while preserving the variance on `S1` and `S2`.

Anyway, completely understand there may not be a resolution in the near future but just one data point that at least for ZIO ecosystem libraries this is an issue we have run into in the wild and is impacting our ability to do something that I think is reasonable and doesn't have a great solution otherwise.

Thanks again for your time and really appreciate you engaging with me on this and all the time from you and the rest of the team on Scala 3!

**Comment by prolativ** [comment]
@adamgfraser I played a bit more with your code and I've found another solution. If you don't have a strict requirement of staying source compatible with scala 2 you can use the new extension method syntax with a little trick to make everything work preserving the 2 desired properties which scala 2 version had.
```scala
object MyLibrary {

  trait ZPure[+W, -S1, +S2, -R, +E, +A]

  object ZPure {
    extension [W, S, R, E, A](self: ZPure[W, S, S, R, E, A])
      def zipPar[B](that: ZPure[W, S, S, R, E, B]): ZPure[W, S, S, R, E, (A, B)] = ???
  }
  
  trait CommutativeBoth[F[+_]] {
    def both[A ,B](fa: F[A], fb: F[B]): F[(A, B)]
  }

  given [A, B]: CommutativeBoth[List] with
    def both[A ,B](fa: List[A], fb: List[B]): List[(A, B)] = ???

  extension [F[+_], A](fa: => F[A])
    def zipPar[B](using both: CommutativeBoth[F])(that: => F[B]): F[(A, B)] = ???
}

object UserCode extends App {
  import MyLibrary.{_, given}
  
  lazy val fooPure: ZPure[Nothing, Unit, Unit, Any, String, Int] = ???
  fooPure zipPar fooPure

  lazy val fooList = List(1, 2, 3)
  fooList zipPar fooList
  
  println("It compiles!")
}
```
Here also `zipPar` defined directly in `MyLibrary` has higher priority than `zipPar` from `ZPure` object but because the `using` clause of the extension proceeds the receiver parameter, if the implicit is not found, the implicit search continues and the proper extension method gets found.
The syntax changes needed for that to work are not there in 3.0.0-M3 but you can try it with the nightly release
